### PR TITLE
#211 Extract sync logic from main.rs

### LINKED ISSUE #211: Improve mechanism for initialising database
Would be good to have a way to initialise the database with suitable mock data. 

Options:

- Insert mock data in code
- Run script to insert mock data via mutations
- Run script/`pg_dump` to directly initialise database

I think for now the last option works best? It's database dependent, but should be fine (unlikely we will be switching out DBs anytime soon!)

### PR DESCRIPTION
**Pull request by wlthomson**
Closes #211.

### CODE REVIEW

File: src/util/sync.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            Err(mpsc_error::TrySendError::Full(())) => {
+                info!("Failed to send sync message as another sync is currently in progress.")
+            }
+            Err(mpsc_error::TrySendError::Closed(())) => {
+                unreachable!("Sync channel has closed. Is the receiver dead!?")
+            }
+        }
+    }
+
+    pub async fn schedule_send(&mut self, interval_duration: Duration) {
```
Comment (andreievg) [inline src/util/sync.rs:34]: I would assume there will be a scheduler that would send sync requests (one implementation for all of the delay/resumes).  Just FYI, I think what you have is good for now

File: src/util/sync.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+pub struct SyncReceiverActor {
+    receiver: MpscReceiver<()>,
+}
+
+impl SyncReceiverActor {
+    pub async fn listen(&mut self, delay_duration: Duration) {
+        while let Some(()) = self.receiver.recv().await {
+            info!("Received sync message. Simulating sync...");
+            time::delay_for(delay_duration).await;
```
Comment (andreievg) [inline src/util/sync.rs:55]: is this delay just to show that sync is going to be doing some work ? in that case I think hard code the delay_duratio (no need for paramters)

**Review by andreievg — COMMENTED** [review]
This is looking very good.

Although would have thought that sync would be a service that is initilise similar to httpserver.

Sync::new(receive_channel).run()

What do you think ?

File: src/util/sync.rs (hunk 2 above)
Comment (wlthomson) [reply src/util/sync.rs:55]: Yeah, it was just to simulate the sync. This is redundant as of #233, though.

**Comment by wlthomson** [comment]
> This is looking very good.
> 
> Although would have thought that sync would be a service that is initilise similar to httpserver.
>
> Sync::new(receive_channel).run()
>
> What do you think ?

Check out #233, might be a bit closer to what you had in mind!

**Comment by andreievg** [comment]
> Check out #233, might be a bit closer to what you had in mind!

@wlthomson this looks like 0 lines of code

**Comment by wlthomson** [comment]
> 
> 
> > Check out #233, might be a bit closer to what you had in mind!
> 
> @wlthomson this looks like 0 lines of code

Haha, sorry! Meant to link [this one](https://github.com/openmsupply/remote-server/pull/239).

**Comment by andreievg** [comment]
Looks good, will approve as soon as pre-requisite PR is approved

**Review by andreievg — APPROVED** [review]

**Pull request merged**